### PR TITLE
Use OBOD report as the source of truth of in-memory locations of object

### LINKED ISSUE #25084: [Core] Owner has incorrect in-memory locations of object
### What happened + What you expected to happen

Current Ray owner reference counter has two channels of updating in-memory locations of an object: one is obod reports from raylet and the other is `ReferenceCounter::UpdateObjectPinnedAtRaylet` which happens when the PushTask returns and when we pin a secondary copy. The issue with two channels is that they are not ordered, so one update can overwrites the other newer update.

One example is that the primary worker node puts the task return value to plasma store and then immediately spills it. In this case, the owner will see `object add to In-memory`, `object remove from In-memory` and `object spilled` from obod channel. Also the owner will see PushTaskReply in the other channel and will call `UpdateObjectPinnedAtRaylet` to add the worker node to the in-memory location. It's possible the owner sees obod reports first and then the PushTaskReply, in this case, object in-memory location will have the worker node even though it only exists on disk.

Another issue is that when the primary copy of an object is lost, owner will try to pin a secondary copy but its' possible that the secondary copy is evicted in the meantime. In this case, the secondary copy raylet should return error instead of ok so owner knows that secondary copy is lost and won't think the pin succeeded.

### Versions / Dependencies

master

### Reproduction script

push-bases shuffle chaos test

### Issue Severity

High: It blocks me from completing my task.

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Currently Owner has two channels of updating in-memory locations of object: OBOD report and `ReferenceCounter::UpdateObjectPinnedAtRaylet`. These two channels are not ordered so one channel may overwrite the other channel. For example `ReferenceCounter::UpdateObjectPinnedAtRaylet` may add a node to the in-memory location after it's removed based on OBOD report.

This PR fixes this by using OBOD reports as the source of truth for the in-memory locations of object. `ReferenceCounter::UpdateObjectPinnedAtRaylet` no longer updates the in-memory locations of object.

## Related issue number
Closes #25084

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by stephanie-wang — COMMENTED** [review]
Nice fixes! However, could you split it into two PRs? This will make it easier to revert if needed.

File: src/ray/protobuf/node_manager.proto [hunk 1]
```diff
@@ -169,7 +169,9 @@ message PinObjectIDsRequest {
   repeated bytes object_ids = 2;
 }
 
-message PinObjectIDsReply {}
+message PinObjectIDsReply {
+  repeated bool success = 1;
```
Comment (scv119) [inline src/ray/protobuf/node_manager.proto:173]: nit: should we use plural here?

File: src/ray/core_worker/reference_count.cc [hunk 2]
```diff
@@ -1281,7 +1279,12 @@ absl::optional<LocalityData> ReferenceCounter::GetLocalityData(
...
-  const auto &node_ids = it->second.locations;
+  auto node_ids = it->second.locations;
+  // Add location of the primary copy since the object must be there: either in memory or
+  // spilled.
+  if (it->second.pinned_at_raylet_id.has_value()) {
+    node_ids.emplace(it->second.pinned_at_raylet_id.value());
+  }
 
   // We should only reach here if we have valid locality data to return.
   absl::optional<LocalityData> locality_data(
```
Comment (scv119) [inline src/ray/core_worker/reference_count.cc:1290]: move the node_ids in?

File: src/ray/protobuf/node_manager.proto (hunk 1 above)
Comment (jjyao) [reply src/ray/protobuf/node_manager.proto:173]: Addressed in https://github.com/ray-project/ray/pull/25104 since I split the PR into two.

**Comment by jjyao** [comment]
Release tests look good: https://buildkite.com/ray-project/release-tests-pr/builds/4073

**Review by stephanie-wang — APPROVED** [review]

**Pull request merged**